arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,3181 @@
+ *  dev  - Reference to the NuttX driver state structure
+ *
+ * Returned Value:
+ *  Zero (OK) on success; a negated errno value on failure
+ *
+ * Assumptions:
+ *  None
+ *
+ ****************************************************************************/
+
+static int mpfs_ifdown(struct net_driver_s *dev)
+{
+  FAR struct mpfs_driver_s *priv =
+    (FAR struct mpfs_driver_s *)dev->d_private;
+
+  /* Stop chip */
+
+  mpfs_can_chip_stop(priv);
+
+  priv->bifup = false;
+  return OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_ioctl
+ *
+ * Description:
+ *  PHY ioctl command handler
+ *
+ * Input Parameters:
+ *  dev  - Reference to the NuttX driver state structure
+ *  cmd  - ioctl command
+ *  arg  - Argument accompanying the command
+ *
+ * Returned Value:
+ *  Zero (OK) on success; a negated errno value on failure.
+ *
+ * Assumptions:
+ *  None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_NETDEV_IOCTL
+static int mpfs_ioctl(struct net_driver_s *dev, int cmd, unsigned long arg)
+{
+#if defined(CONFIG_NETDEV_CAN_BITRATE_IOCTL) || \
+defined(CONFIG_NETDEV_CAN_FILTER_IOCTL)
+  FAR struct mpfs_driver_s *priv =
+    (FAR struct mpfs_driver_s *)dev->d_private;
+#endif
+  int ret;
+
+  switch (cmd)
+    {
+#ifdef CONFIG_NETDEV_CAN_BITRATE_IOCTL
+    case SIOCGCANBITRATE:
+
+      /* Get bitrate from the FPGA CANFD controller */
+
+      {
+        struct can_ioctl_data_s *req =
+          (struct can_ioctl_data_s *)((uintptr_t)arg);
+        req->arbi_bitrate = priv->can.bittiming.bitrate / 1000; /* kbit/s */
+        req->arbi_samplep = priv->can.bittiming.sample_point / 10;
+        req->data_bitrate = priv->can.data_bittiming.bitrate / 1000; /* kbit/s */
+        req->data_samplep = priv->can.data_bittiming.sample_point / 10;
+        ret = OK;
+      }
+      break;
+
+    case SIOCSCANBITRATE:
+
+      /* Set bitrate of the FPGA CANFD controller */
+
+      {
+        struct can_ioctl_data_s *req =
+          (struct can_ioctl_data_s *)((uintptr_t)arg);
+
+        if (req->arbi_bitrate > 1000)
+          {
+            canerr("Arbitration bitrate > 1Mbps is not supported.");
+            ret = -EINVAL;
+            break;
+          }
+        priv->can.bittiming.bitrate = req->arbi_bitrate * 1000; /* bit/s */
+
+        if (req->arbi_samplep > 100 || req->arbi_samplep <= 0)
+          {
+            canerr("Invalid arbitration sample point. "
+                   "Range should be (0,100]%%.");
+            ret = -EINVAL;
+            break;
+          }
+        priv->can.bittiming.sample_point =
+          req->arbi_samplep * 10; /* In one-tenth of a percent */
+
+        if (req->data_bitrate > 4000 ||
+            req->data_bitrate < req->arbi_bitrate)
+          {
+            canerr("Data bitrate higher than 4Mbps is yet to be supported. "
+                   "Data br cannot be smaller than arbitration br.");
```

Comment:
Why not? (Does not make sense as SNR for arbitration will be better, but why is this not possible)?